File: sdk/perl/benchmarks/client_benchmark.pl
#!/usr/bin/env perl
use v5.32;
use strict;
use warnings;
use FindBin;
use Getopt::Long qw(GetOptions);
use Time::HiRes qw(time);

use lib "$FindBin::Bin/../lib";
use GlyphaStore;
use GlyphaStore::Client;

my %options = (
    host     => '127.0.0.1',
    workers  => 1,
    ops      => 100_000,
    pipeline => 128,
    warmup   => 1,
    repeats  => 7,
);
GetOptions(
    'host=s'     => \$options{host},
    'port=i'     => \$options{port},
    'workers=i'  => \$options{workers},
    'ops=i'      => \$options{ops},
    'pipeline=i' => \$options{pipeline},
    'warmup=i'   => \$options{warmup},
    'repeats=i'  => \$options{repeats},
) or die "invalid benchmark arguments\n";
die "--port is required\n" if !$options{port};
die "numeric arguments are outside benchmark limits\n"
    if $options{workers} < 1 || $options{ops} < 1 || $options{pipeline} < 1
    || $options{warmup} < 0 || $options{repeats} < 1;

my @requests = map { [] } 1 .. $options{workers};
my @remaining = map {
    int($options{ops} / $options{workers})
        + ($_ < $options{ops} % $options{workers} ? 1 : 0)
} 0 .. $options{workers} - 1;
my $candidate = 0;
while (grep { $_ } @remaining) {
    my $key = sprintf('perl-bench-%012d', $candidate);
    my $owner = GlyphaStore::Protocol::worker_for($key, $options{workers});
    if ($remaining[$owner]) {
        my $value = chr($candidate & 0xFF) x 64;
        push @{$requests[$owner]}, { opcode => 'put', key => $key, value => $value };
        push @{$requests[$owner]}, { opcode => 'get', key => $key };
        --$remaining[$owner];
    }
    ++$candidate;
}
my $batch_frames = $options{pipeline} * 2;
my @batches;
for my $worker_requests (@requests) {
    my @worker_batches;
    for (my $offset = 0; $offset < @$worker_requests; $offset += $batch_frames) {
        my $last = $offset + $batch_frames - 1;
        $last = $#$worker_requests if $last > $#$worker_requests;
        push @worker_batches, [@$worker_requests[$offset .. $last]];
    }
    push @batches, \@worker_batches;
}

my $client = GlyphaStore::Client->connect(host => $options{host}, port => $options{port});
die "server Worker count does not match --workers\n"
    if $client->worker_count != $options{workers};

sub run_once {
    my $started = time;
    for my $worker_batches (@batches) {
        for my $batch (@$worker_batches) {
            my $responses = $client->execute_pipeline($batch);
            die "pipeline response count mismatch\n" if @$responses != @$batch;
            for my $index (0 .. $#$batch) {
                die "pipeline request failed\n"
                    if $responses->[$index]->{outcome} ne 'succeeded';
                if ($batch->[$index]->{opcode} eq 'get') {
                    die "pipeline GET value mismatch\n"
                        if $responses->[$index]->{value} ne $batch->[$index - 1]->{value};
                }
            }
        }
    }
    return time - $started;
}

sub median {
    my @sorted = sort { $a <=> $b } @_;
    return $sorted[int(@sorted / 2)] if @sorted % 2;
    return ($sorted[@sorted / 2 - 1] + $sorted[@sorted / 2]) / 2;
}

run_once() for 1 .. $options{warmup};
my @samples = map { run_once() } 1 .. $options{repeats};
$client->close;

my $sdk_version = $GlyphaStore::VERSION;
my $operation_count = $options{ops} * 2;
my @rates = map { $operation_count / $_ } @samples;
printf "# glyphastore Perl client benchmark\n";
printf "# sdk_version=%s runtime=sync execution=single-process-worker-sequential "
    . "workers=%d pipeline_pairs=%d operations=%d\n",
    $sdk_version, $options{workers}, $options{pipeline}, $operation_count;
printf "name=perl_client_pipeline_read_after_write sdk_version=%s runtime=sync "
    . "execution=single-process-worker-sequential workers=%d pipeline_pairs=%d "
    . "operations=%d samples=%d median_seconds=%.9f min_seconds=%.9f max_seconds=%.9f "
    . "median_ops_per_second=%.3f min_ops_per_second=%.3f max_ops_per_second=%.3f\n",
    $sdk_version, $options{workers}, $options{pipeline}, $operation_count,
    scalar(@samples), median(@samples), (sort { $a <=> $b } @samples)[0],
    (sort { $b <=> $a } @samples)[0], median(@rates),
    (sort { $a <=> $b } @rates)[0], (sort { $b <=> $a } @rates)[0];

File: sdk/ruby/benchmarks/client_benchmark.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "optparse"
require "pathname"
require_relative "../lib/glypha_store"

options = {
  host: "127.0.0.1",
  workers: 1,
  ops: 100_000,
  pipeline: 128,
  warmup: 1,
  repeats: 7,
  concurrent: nil
}
parser = OptionParser.new do |opts|
  opts.banner = "Usage: client_benchmark.rb --port N [options]"
  opts.on("--host HOST", String) { |v| options[:host] = v }
  opts.on("--port PORT", Integer) { |v| options[:port] = v }
  opts.on("--workers N", Integer) { |v| options[:workers] = v }
  opts.on("--ops N", Integer) { |v| options[:ops] = v }
  opts.on("--pipeline N", Integer) { |v| options[:pipeline] = v }
  opts.on("--warmup N", Integer) { |v| options[:warmup] = v }
  opts.on("--repeats N", Integer) { |v| options[:repeats] = v }
  opts.on("--[no-]concurrent") { |v| options[:concurrent] = v }
end
parser.parse!
abort("--port is required") if options[:port].nil?
abort("numeric arguments are outside benchmark limits") if options[:workers] < 1 || options[:ops] < 1 ||
  options[:pipeline] < 1 || options[:warmup].negative? || options[:repeats] < 1

use_concurrent = if options[:concurrent].nil?
                   options[:workers] > 1
                 else
                   options[:concurrent]
                 end

requests = Array.new(options[:workers]) { [] }
remaining = options[:workers].times.map do |i|
  (options[:ops] / options[:workers]) + (i < (options[:ops] % options[:workers]) ? 1 : 0)
end
candidate = 0
while remaining.any?(&:positive?)
  key = format("ruby-bench-%012d", candidate).b
  owner = GlyphaStore::Protocol.worker_for(key, options[:workers])
  if remaining[owner].positive?
    value = ([candidate & 0xFF].pack("C") * 64)
    requests[owner] << GlyphaStore::PipelineRequest.new(
      opcode: GlyphaStore::PipelineOpcode::PUT, key: key, value: value
    )
    requests[owner] << GlyphaStore::PipelineRequest.new(
      opcode: GlyphaStore::PipelineOpcode::GET, key: key
    )
    remaining[owner] -= 1
  end
  candidate += 1
end

batch_frames = options[:pipeline] * 2
batches = []
max_rounds = 0
requests.each do |worker_requests|
  worker_batches = []
  offset = 0
  while offset < worker_requests.length
    last = [offset + batch_frames - 1, worker_requests.length - 1].min
    worker_batches << worker_requests[offset..last]
    offset = last + 1
  end
  max_rounds = worker_batches.length if worker_batches.length > max_rounds
  batches << worker_batches
end

config = GlyphaStore::ClientConfig.defaults
config.host = options[:host]
config.port = options[:port]
config.maximum_pipeline_requests = batch_frames
client = GlyphaStore::Client.connect(config)
abort("server Worker count does not match --workers") if client.worker_count != options[:workers]

def validate_batch!(batch, responses)
  raise "pipeline response count mismatch" if responses.length != batch.length

  batch.each_with_index do |request, index|
    raise "pipeline request failed" unless responses[index].succeeded?

    next unless request.opcode == GlyphaStore::PipelineOpcode::GET
    raise "pipeline GET value mismatch" if responses[index].value != batch[index - 1].value
  end
end

run_once =
  if use_concurrent
    lambda do
      started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      max_rounds.times do |round|
        wave = batches.map { |worker_batches| round < worker_batches.length ? worker_batches[round] : [] }
        # Concurrent: one pipeline per Worker via execute_batch flattened... use threads like batch.
        threads = []
        results = Array.new(options[:workers])
        options[:workers].times do |worker|
          next if wave[worker].empty?

          threads << Thread.new(worker) do |w|
            results[w] = client.execute_pipeline(wave[w])
          end
        end
        threads.each(&:join)
        options[:workers].times do |worker|
          next if wave[worker].empty?

          validate_batch!(wave[worker], results[worker])
        end
      end
      Process.clock_gettime(Process::CLOCK_MONOTONIC) - started
    end
  else
    lambda do
      started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      batches.each do |worker_batches|
        worker_batches.each do |batch|
          validate_batch!(batch, client.execute_pipeline(batch))
        end
      end
      Process.clock_gettime(Process::CLOCK_MONOTONIC) - started
    end
  end

options[:warmup].times { run_once.call }
samples = options[:repeats].times.map { run_once.call }
client.close

operation_count = options[:ops] * 2
rates = samples.map { |s| operation_count / s }
sorted = samples.sort
median_s = sorted.length.odd? ? sorted[sorted.length / 2] : (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2.0
sorted_rates = rates.sort
median_r = sorted_rates.length.odd? ? sorted_rates[sorted_rates.length / 2] : (sorted_rates[sorted_rates.length / 2 - 1] + sorted_rates[sorted_rates.length / 2]) / 2.0
execution = use_concurrent ? "single-process-worker-concurrent" : "single-process-worker-sequential"

puts "# glyphastore Ruby client benchmark"
puts "# sdk_version=#{GlyphaStore::VERSION} runtime=sync execution=#{execution} " \
     "workers=#{options[:workers]} pipeline_pairs=#{options[:pipeline]} operations=#{operation_count}"
printf(
  "name=ruby_client_pipeline_read_after_write sdk_version=%s runtime=sync execution=%s " \
  "workers=%d pipeline_pairs=%d operations=%d samples=%d median_seconds=%.9f min_seconds=%.9f " \
  "max_seconds=%.9f median_ops_per_second=%.3f min_ops_per_second=%.3f max_ops_per_second=%.3f\n",
  GlyphaStore::VERSION, execution, options[:workers], options[:pipeline], operation_count,
  samples.length, median_s, sorted.first, sorted.last, median_r, sorted_rates.first, sorted_rates.last
)
